feat(ui): add 'smoothscroll' option

luukvbaal opened the pull request:
Fix #22489
Close #8715
Close #11014 (superseded)

**vim-patch:9.0.0640: cannot scroll by screen line if a line wraps**

Problem:    Cannot scroll by screen line if a line wraps.
Solution:   Add the 'smoothscroll' option.  Only works for CTRL-E and CTRL-Y
            so far.

https://github.com/vim/vim/commit/f6196f424474e2a9c160f2a995fc2691f82b58f9

**vim-patch:9.0.0641: missing part of the new option code**

Problem:    Missing part of the new option code.
Solution:   Add missing WV_SMS.

https://github.com/vim/vim/commit/bbbda8fd81f6d720962b67ae885825bad9be4456

**vim-patch:9.0.0642: breakindent test fails**

Problem:    Breakindent test fails.
Solution:   Correct logic for resetting need_showbreak.

https://github.com/vim/vim/commit/693729ae58bd30fc1a4c08042ebe9923b45f5763

**vim-patch:9.0.0643: smoothscroll test fails**

Problem:    Smoothscroll test fails.
Solution:   Check if skipcol changed.

https://github.com/vim/vim/commit/b34c4b7863af8718ad726173585dd38a7c292f0f

**vim-patch:9.0.0644: 'smoothscroll' is not copied to a new window on :split**

Problem:    'smoothscroll' is not copied to a new window on :split.
Solution:   Copy the option value.  Add a test.

https://github.com/vim/vim/commit/b1fd26d208aadc96d3e8b9215f761150f40a9f91

**vim-patch:9.0.0645: CTRL-Y does not stop at line 1**

Problem:    CTRL-Y does not stop at line 1. (John Marriott)
Solution:   Stop at line 1 when 'smoothscroll' is not set. (closes vim/vim#11261)

https://github.com/vim/vim/commit/8df9748edb2ac8bd025e34e06194ac210667c97a

**vim-patch:9.0.0646: with 'smoothscroll' CTRL-E is wrong when 'foldmethod' set**

Problem:    with 'smoothscroll' set CTRL-E does not work properly when
            'foldmethod' is set to "indent". (Yee Cheng Chin)
Solution:   Merge the code for scroling with folds and 'smoothscroll'.
            (closes vim/vim#11262)

https://github.com/vim/vim/commit/6b2d4ff7148e0b416ba745d20d061e6f7bb53ee7

**vim-patch:9.0.0649: no indication the first line is broken for 'smoothscroll'**

Problem:    No indication when the first line is broken for 'smoothscroll'.
Solution:   Show "<<<" in the first line.

https://github.com/vim/vim/commit/406b5d89e18742ac6e6256ffc72fb70a27f0148b


**vim-patch:9.0.0650: some tests are failing**

Problem:    Some tests are failing.
Solution:   Adjust for "<<<" showing up.

https://github.com/vim/vim/commit/0466d398a550623126fba3a2a08208a798b28bda

**vim-patch:9.0.0652: 'smoothscroll' not tested with 'number' and "n" in 'cpo'**

Problem:    'smoothscroll' not tested with 'number' and "n" in 'cpo'.
Solution:   Add tests, fix uncovered problem.

https://github.com/vim/vim/commit/b6aab8f44beb8c5d99393abdc2c9faab085c72aa

**vim-patch:9.0.0654: breakindent test fails**

Problem:    Breakindent test fails.
Solution:   Temporarily accept wrong result.

https://github.com/vim/vim/commit/ec32c781a282398e3da27f4aec4b03fcd20f8b0d

**vim-patch:9.0.0671: negative topline using CTRL-Y with 'smoothscroll' and 'diff'**

Problem:    Negative topline using CTRL-Y with 'smoothscroll' and 'diff'.
                (Ernie Rael)
Solution:   Only use 'smoothscroll' when 'wrap' is set.

https://github.com/vim/vim/commit/1a58e1d97cfc72e501cbf63ad75f46f1bb4c8da2

**vim-patch:9.0.0672: line partly shows with 'smoothscroll' and 'scrolloff' zero**

Problem:    Cursor line only partly shows with 'smoothscroll' and 'scrolloff'
            zero.
Solution:   Do not use 'smoothscroll' when adjusting the bottom of the window.
            (closes vim/vim#11269)

https://github.com/vim/vim/commit/9bab7a024393200bb2c03b3abddfda86436990a7

**vim-patch:9.0.0673: first line wong with 'smoothscroll' and 'scrolloff' zero**

Problem:    First line not scrolled properly with 'smoothscroll' and
            'scrolloff' zero and using "k".
Solution:   Make sure the cursor position is visible.

https://github.com/vim/vim/commit/46b54747c5d252c584571a321231bad9330018ec

**vim-patch:9.0.0677: breakindent test accepts wrong result**

Problem:    Breakindent test accepts wrong result.
Solution:   Fix the number column and adjust the expected text.

https://github.com/vim/vim/commit/06618f94f1c9ed73a84ad5d6a8e1933b0b8da846

**vim-patch:9.0.0679: tests failing with 'smoothscroll', 'number' and "n" in 'cpo'**

Problem:    Tests failing with 'smoothscroll', 'number' and "n" in 'cpo'.
Solution:   Do not count number column in topline if columns are skipped.

https://github.com/vim/vim/commit/35b251d2c2c9d415887d334473669ea886117356

**vim-patch:9.0.0680: tests failing with 'breakindent', 'number' and "n" in 'cpo'**

Problem:    Tests failing with 'breakindent', 'number' and "n" in 'cpo'.
Solution:   Do count the number column in topline if 'breakindent' is set.

https://github.com/vim/vim/commit/3725116f6ec3b5c01e456b151a60c0690e04f76c


**vim-patch:9.0.0681: "<<<" shows for 'smoothscroll' even when 'showbreak is set**

Problem:    "<<<" shows for 'smoothscroll' even when 'showbreak is
set.
Solution:   When 'showbreak' is set do not display "<<<".

https://github.com/vim/vim/commit/0937b9fb244949b7ce9bfcf8398d7495b9b6aa85

**vim-patch:9.0.0701: with 'smoothscroll' cursor position not adjusted in long line**

Problem:    With 'smoothscroll' the cursor position s not adjusted in a long
            line.
Solution:   Move the cursor further up or down in the line.

https://github.com/vim/vim/commit/8cf3459878198c5bb4a96f3c63214b2beccce341

**vim-patch:9.0.0707: with 'smoothscroll' cursor position not adjusted in long line**

Problem:    With 'smoothscroll' and 'scrolloff' non-zero the cursor position
            is not properly adjusted in a long line.
Solution:   Move the cursor further up or down in the line.

https://github.com/vim/vim/commit/118c235112854f34182d968613d7fe98be3b290b

**vim-patch:9.0.0734: cursor position invalid when scrolling with 'smoothscroll'**

Problem:    Cursor position invalid when scrolling with 'smoothscroll' set.
            (Ernie Rael)
Solution:   Add w_valid_skipcol and clear flags when it changes.  Adjust
            w_skipcol after moving the cursor.

https://github.com/vim/vim/commit/2fbabd238a94022c99506e920186a5b6cdf15426

**vim-patch:9.0.0735: breakindent and scrolloff tests fail**

Problem:    Breakindent and scrolloff tests fail.
Solution:   Temporarily skip the assertions.

https://github.com/vim/vim/commit/e42033e735febb163fdc57aadbc11787ca611ba9

**vim-patch:9.0.0745: wrong cursor position when using "gj" and "gk" in a long line**

Problem:    Wrong cursor position when using "gj" and "gk" in a long line.
Solution:   Adjust computations for the cursor position and skipcol. Re-enable
            tests that pass now, disable failing breakindent test.

https://github.com/vim/vim/commit/4b6172e108fe06be47c09f8690dc54608be3ee80

**vim-patch:9.0.0746: breakindent test cases are commented out**

Problem:    Breakindent test cases are commented out.
Solution:   Adjust expected result to slightly different behavior.  Correct
            computations for cursor position.

https://github.com/vim/vim/commit/856c5d2bc7c3864f8b63a0ab3e376d5c5e51f1d5

**vim-patch:9.0.0751: 'scrolloff' does not work well with 'smoothscroll'**

Problem:    'scrolloff' does not work well with 'smoothscroll'.
Solution:   Make positioning the cursor a bit better.  Rename functions.

https://github.com/vim/vim/commit/c9121f798f49fa71e814912cb186d89c164090c3

**vim-patch:9.0.0757: line number not visisble with 'smoothscroll', 'nu' and 'rnu'**

Problem:    Line number not visisble with 'smoothscroll', 'nu' and 'rnu'.
Solution:   Put the ">>>" after the line number instead of on top.

https://github.com/vim/vim/commit/eb4de629315f2682d8b314462d02422ec98d751a

**vim-patch:9.0.0758: "precedes" from 'listchars' overwritten by <<<**

Problem:    "precedes" from 'listchars' overwritten by <<< for 'smoothscroll'.
Solution:   Keep the "precedes" character.

https://github.com/vim/vim/commit/13cdde39520220bb856cba16626327c706752b51

**vim-patch:9.0.0760: display test for 'listchars' "precedes" fails**

Problem:    Display test for 'listchars' "precedes" fails.
Solution:   Correct the expected result.

https://github.com/vim/vim/commit/297164cb7972beff35e375ccac4fbad8196ccbd7

**vim-patch:9.0.0807: with 'smoothscroll' typing "0" may not go to the first column**

Problem:    With 'smoothscroll' typing "0" may not go to the first column.
Solution:   Recompute w_cline_height when needed.  Do not scroll up when it
            would move the cursor.

https://github.com/vim/vim/commit/d5337efece7c68e9b4ce864532ea49b02453b674

**vim-patch:9.0.0892: may redraw when not needed**

Problem:    May redraw when not needed, causing slow scrolling.
Solution:   Do not redraw when w_skipcol doesn't change.  When w_skipcol
            changes only redraw from the top. (issue vim/vim#11559)

https://github.com/vim/vim/commit/f32fb93e431e4db95a8663d86dfeb6bffa5896f6

**vim-patch:9.0.0893: 'smoothscroll' cursor calculations wrong when 'number' is set**

Problem:    'smoothscroll' cursor calculations wrong when 'number' is set.
Solution:   Correct the code that computes the width. (closes vim/vim#11492)

https://github.com/vim/vim/commit/01ee52bab6041450095c53f9469b1b266a7e3d4d

**vim-patch:9.0.0898: with 'smoothscroll' cursor is one screen line too far down**

Problem:    With 'smoothscroll' cursor is one screen line too far down. (Ernie
            Rael)
Solution:   Add a test that currently has the wrong result so that a fix can
            be made. (issue vim/vim#11436)

https://github.com/vim/vim/commit/75ac25b4967cdcdfdf2d6c086a6e2308868c280a

**vim-patch:9.0.0900: cursor moves too far with 'smoothscroll'**

Problem:    Cursor moves too far with 'smoothscroll'.
Solution:   Only move as far as really needed. (Yee Cheng Chin, closes vim/vim#11504)

https://github.com/vim/vim/commit/81ba26e9de24ca6b1c05b6ec03e53b21793f1a4b

**vim-patch:9.0.0901: setting w_leftcol and handling side effects is confusing**

Problem:    Setting w_leftcol and handling side effects is confusing.
Solution:   Use a function to set w_leftcol() and handle side effects.

https://github.com/vim/vim/commit/0c34d562647f029faca40f7733ccfb7b5377672b

**vim-patch:9.0.0908: with 'smoothscroll' cursor may end up in wrong position**

Problem:    With 'smoothscroll' cursor may end up in wrong position.
Solution:   Correct the computation of screen lines. (Yee Cheng Chin,
            closes vim/vim#11502)

https://github.com/vim/vim/commit/361895d2a15b4b0bbbb4c009261eab5b3d69ebf1

**vim-patch:9.0.0911: with 'smoothscroll' set mouse click position may be wrong**

Problem:    With 'smoothscroll' set mouse click position may be wrong.
Solution:   Adjust computations for w_skipcol. (Yee Cheng Chin, closes vim/vim#11514)

https://github.com/vim/vim/commit/e6392b102151ec69fad232bcf00591230cef8e1c

**vim-patch:9.0.0998: "gk" may reset skipcol when not needed**

Problem:    "gk" may reset skipcol when not needed.
Solution:   Only reset skipcol if the cursor column is less.

https://github.com/vim/vim/commit/1b73edd9ee40aec400f3611f59823cec5fd1c489

**vim-patch:9.0.1000: with 'smoothscroll' skipcol may be reset unnecessarily**

Problem:    With 'smoothscroll' skipcol may be reset unnecessarily.
Solution:   Check the line does actually fit in the window.

https://github.com/vim/vim/commit/b21b8e9ed081a6ef6b6745fe65d219b3ac046c3b

**vim-patch:9.0.1121: cursor positioning and display problems with 'smoothscroll'**

Problem:    Cursor positioning and display problems with 'smoothscroll' and
            using "zt", "zb" or "zz".
Solution:   Adjust computations and conditions. (Yee Cheng Chin,
            closes vim/vim#11764)

https://github.com/vim/vim/commit/db4d88c2adfe8f8122341ac9d6cae27ef78451c8

**vim-patch:9.0.1247: divide by zero with 'smoothscroll' set and a narrow window**

Problem:    Divide by zero with 'smoothscroll' set and a narrow window.
Solution:   Bail out when the window is too narrow.

https://github.com/vim/vim/commit/870219c58c0804bdc55419b2e455c06ac715a835

**vim-patch:9.0.1367: divide by zero in zero-width window**

Problem:    Divide by zero in zero-width window.
Solution:   Check the width is positive.

https://github.com/vim/vim/commit/e0f869196930ef5f25a0ac41c9215b09c9ce2d3c

**vim-patch:9.0.1435: scrolling too many lines when 'wrap' and 'diff' are set**

Problem:    Scrolling too many lines when 'wrap' and 'diff' are set.
Solution:   Only scroll by screenlines for 'diff' when 'wrap' is not set.
            (closes vim/vim#12211)

https://github.com/vim/vim/commit/38d867f041349e1400c2cce9cac06f59ae6ccbb1

**vim-patch:9.0.1502: no test for deleting the end of a long wrapped line**

Problem:    No test for deleting the end of a long wrapped line.
Solution:   Add a test to check the right text is displayed. (Luuk van Baal,
            closes vim/vim#12318)

https://github.com/vim/vim/commit/5b10a140983c16140d69a214494c4b8af8b34763

luukvbaal committed:
vim-patch:9.0.0640: cannot scroll by screen line if a line wraps

Problem:    Cannot scroll by screen line if a line wraps.
Solution:   Add the 'smoothscroll' option.  Only works for CTRL-E and CTRL-Y
            so far.

https://github.com/vim/vim/commit/f6196f424474e2a9c160f2a995fc2691f82b58f9

vim-patch:9.0.0641: missing part of the new option code

Problem:    Missing part of the new option code.
Solution:   Add missing WV_SMS.

https://github.com/vim/vim/commit/bbbda8fd81f6d720962b67ae885825bad9be4456

Co-authored-by: Bram Moolenaar <Bram@vim.org>

zeertzjq commented:
Probably also fixes #22489

luukvbaal committed:
vim-patch:9.0.0642: breakindent test fails

Problem:    Breakindent test fails.
Solution:   Correct logic for resetting need_showbreak.

https://github.com/vim/vim/commit/693729ae58bd30fc1a4c08042ebe9923b45f5763

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0643: smoothscroll test fails

Problem:    Smoothscroll test fails.
Solution:   Check if skipcol changed.

https://github.com/vim/vim/commit/b34c4b7863af8718ad726173585dd38a7c292f0f

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0644: 'smoothscroll' is not copied to a new window on :split

Problem:    'smoothscroll' is not copied to a new window on :split.
Solution:   Copy the option value.  Add a test.

https://github.com/vim/vim/commit/b1fd26d208aadc96d3e8b9215f761150f40a9f91

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0645: CTRL-Y does not stop at line 1

Problem:    CTRL-Y does not stop at line 1. (John Marriott)
Solution:   Stop at line 1 when 'smoothscroll' is not set. (closes vim/vim#11261)

https://github.com/vim/vim/commit/8df9748edb2ac8bd025e34e06194ac210667c97a

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0646: with 'smoothscroll' CTRL-E is wrong when 'foldmethod' set

Problem:    with 'smoothscroll' set CTRL-E does not work properly when
            'foldmethod' is set to "indent". (Yee Cheng Chin)
Solution:   Merge the code for scroling with folds and 'smoothscroll'.
            (closes vim/vim#11262)

https://github.com/vim/vim/commit/6b2d4ff7148e0b416ba745d20d061e6f7bb53ee7

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal commented on test/functional/legacy/scroll_opt_spec.lua:
```diff
@@ -0,0 +1,134 @@
+    screen:expect(s1)
+    feed('<C-E>')
+    screen:expect(s2)
+    feed('5G<C-Y>')
```
Not sure where this cursor discrepancy comes from, the oldtest does `7G`. Not sure it matters either.

luukvbaal replied:
Ah it comes from `'scrolloff'`. Do we match the vim test by setting scrolloff to 5? Or leave it like this.

zeertzjq replied:
I think yes.

luukvbaal commented:
Oops, didn't notice that [that](https://github.com/neovim/neovim/actions/runs/4804634761/jobs/8550341910?pr=23320#step:14:5468) test was a vim9 test. What do we do with it?

zeertzjq commented:
> Oops, didn't notice that [that](https://github.com/neovim/neovim/actions/runs/4804634761/jobs/8550341910?pr=23320#step:14:5468) test was a vim9 test. What do we do with it?
> 
> Comment it out and port it to lua? Or port the oldtest to regular vimscript and keep it there?

@justinmk Thoughts?

luukvbaal committed:
vim-patch:9.0.0649: no indication the first line is broken for 'smoothscroll'

Problem:    No indication when the first line is broken for 'smoothscroll'.
Solution:   Show "<<<" in the first line.

https://github.com/vim/vim/commit/406b5d89e18742ac6e6256ffc72fb70a27f0148b

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal commented:
I ported https://github.com/vim/vim/commit/406b5d89e18742ac6e6256ffc72fb70a27f0148b to `grid_put_linebuf()`. Not sure if it should be marked partial.

luukvbaal commented:
> Isn't that exactly what we have vim9jit for?

Porting it to a neovim compatible vimscript test in the same file might make future updates to that test more straightforward to port.

luukvbaal committed:
vim-patch:9.0.0654: breakindent test fails

Problem:    Breakindent test fails.
Solution:   Temporarily accept wrong result.

https://github.com/vim/vim/commit/ec32c781a282398e3da27f4aec4b03fcd20f8b0d

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0671: negative topline using CTRL-Y with 'smoothscroll' and 'diff'

Problem:    Negative topline using CTRL-Y with 'smoothscroll' and 'diff'.
            (Ernie Rael)
Solution:   Only use 'smoothscroll' when 'wrap' is set.

https://github.com/vim/vim/commit/1a58e1d97cfc72e501cbf63ad75f46f1bb4c8da2

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0672: line partly shows with 'smoothscroll' and 'scrolloff' zero

Problem:    Cursor line only partly shows with 'smoothscroll' and 'scrolloff'
            zero.
Solution:   Do not use 'smoothscroll' when adjusting the bottom of the window.
            (closes vim/vim#11269)

https://github.com/vim/vim/commit/9bab7a024393200bb2c03b3abddfda86436990a7

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0673: first line wong with 'smoothscroll' and 'scrolloff' zero

Problem:    First line not scrolled properly with 'smoothscroll' and
            'scrolloff' zero and using "k".
Solution:   Make sure the cursor position is visible.

https://github.com/vim/vim/commit/46b54747c5d252c584571a321231bad9330018ec

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0677: breakindent test accepts wrong result

Problem:    Breakindent test accepts wrong result.
Solution:   Fix the number column and adjust the expected text.

https://github.com/vim/vim/commit/06618f94f1c9ed73a84ad5d6a8e1933b0b8da846

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0679: tests failing with 'smoothscroll', 'number' and "n" in 'cpo'

Problem:    Tests failing with 'smoothscroll', 'number' and "n" in 'cpo'.
Solution:   Do not count number column in topline if columns are skipped.

https://github.com/vim/vim/commit/35b251d2c2c9d415887d334473669ea886117356

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0680: tests failing with 'breakindent', 'number' and "n" in 'cpo'

Problem:    Tests failing with 'breakindent', 'number' and "n" in 'cpo'.
Solution:   Do count the number column in topline if 'breakindent' is set.

https://github.com/vim/vim/commit/3725116f6ec3b5c01e456b151a60c0690e04f76c

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0681: "<<<" shows for 'smoothscroll' even when 'showbreak is set

Problem:    "<<<" shows for 'smoothscroll' even when 'showbreak is set.
Solution:   When 'showbreak' is set do not display "<<<".

https://github.com/vim/vim/commit/0937b9fb244949b7ce9bfcf8398d7495b9b6aa85

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0701: with 'smoothscroll' cursor position not adjusted in long line

Problem:    With 'smoothscroll' the cursor position s not adjusted in a long
            line.
Solution:   Move the cursor further up or down in the line.

https://github.com/vim/vim/commit/8cf3459878198c5bb4a96f3c63214b2beccce341

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0707: with 'smoothscroll' cursor position not adjusted in long line

Problem:    With 'smoothscroll' and 'scrolloff' non-zero the cursor position
            is not properly adjusted in a long line.
Solution:   Move the cursor further up or down in the line.

https://github.com/vim/vim/commit/118c235112854f34182d968613d7fe98be3b290b

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0734: cursor position invalid when scrolling with 'smoothscroll'

Problem:    Cursor position invalid when scrolling with 'smoothscroll' set.
            (Ernie Rael)
Solution:   Add w_valid_skipcol and clear flags when it changes.  Adjust
            w_skipcol after moving the cursor.

https://github.com/vim/vim/commit/2fbabd238a94022c99506e920186a5b6cdf15426

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0735: breakindent and scrolloff tests fail

Problem:    Breakindent and scrolloff tests fail.
Solution:   Temporarily skip the assertions.

https://github.com/vim/vim/commit/e42033e735febb163fdc57aadbc11787ca611ba9

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0745: wrong cursor position when using "gj" and "gk" in a long line

Problem:    Wrong cursor position when using "gj" and "gk" in a long line.
Solution:   Adjust computations for the cursor position and skipcol. Re-enable
            tests that pass now, disable failing breakindent test.

https://github.com/vim/vim/commit/4b6172e108fe06be47c09f8690dc54608be3ee80

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0746: breakindent test cases are commented out

Problem:    Breakindent test cases are commented out.
Solution:   Adjust expected result to slightly different behavior.  Correct
            computations for cursor position.

https://github.com/vim/vim/commit/856c5d2bc7c3864f8b63a0ab3e376d5c5e51f1d5

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0751: 'scrolloff' does not work well with 'smoothscroll'

Problem:    'scrolloff' does not work well with 'smoothscroll'.
Solution:   Make positioning the cursor a bit better.  Rename functions.

https://github.com/vim/vim/commit/c9121f798f49fa71e814912cb186d89c164090c3

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0757: line number not visisble with 'smoothscroll', 'nu' and 'rnu'

Problem:    Line number not visisble with 'smoothscroll', 'nu' and 'rnu'.
Solution:   Put the ">>>" after the line number instead of on top.

https://github.com/vim/vim/commit/eb4de629315f2682d8b314462d02422ec98d751a

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0758: "precedes" from 'listchars' overwritten by <<<

Problem:    "precedes" from 'listchars' overwritten by <<< for 'smoothscroll'.
Solution:   Keep the "precedes" character.

https://github.com/vim/vim/commit/13cdde39520220bb856cba16626327c706752b51

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0760: display test for 'listchars' "precedes" fails

Problem:    Display test for 'listchars' "precedes" fails.
Solution:   Correct the expected result.

https://github.com/vim/vim/commit/297164cb7972beff35e375ccac4fbad8196ccbd7

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0807: with 'smoothscroll' typing "0" may not go to the first column

Problem:    With 'smoothscroll' typing "0" may not go to the first column.
Solution:   Recompute w_cline_height when needed.  Do not scroll up when it
            would move the cursor.

https://github.com/vim/vim/commit/d5337efece7c68e9b4ce864532ea49b02453b674

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0650: some tests are failing

Problem:    Some tests are failing.
Solution:   Adjust for "<<<" showing up.

https://github.com/vim/vim/commit/0466d398a550623126fba3a2a08208a798b28bda

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0652: 'smoothscroll' not tested with 'number' and "n" in 'cpo'

Problem:    'smoothscroll' not tested with 'number' and "n" in 'cpo'.
Solution:   Add tests, fix uncovered problem.

https://github.com/vim/vim/commit/b6aab8f44beb8c5d99393abdc2c9faab085c72aa

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0892: may redraw when not needed

Problem:    May redraw when not needed, causing slow scrolling.
Solution:   Do not redraw when w_skipcol doesn't change.  When w_skipcol
            changes only redraw from the top. (issue vim/vim#11559)

https://github.com/vim/vim/commit/f32fb93e431e4db95a8663d86dfeb6bffa5896f6

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0893: 'smoothscroll' cursor calculations wrong when 'number' is set

Problem:    'smoothscroll' cursor calculations wrong when 'number' is set.
Solution:   Correct the code that computes the width. (closes vim/vim#11492)

https://github.com/vim/vim/commit/01ee52bab6041450095c53f9469b1b266a7e3d4d

Co-authored-by: Yee Cheng Chin <ychin.git@gmail.com>

luukvbaal committed:
vim-patch:9.0.0898: with 'smoothscroll' cursor is one screen line too far down

Problem:    With 'smoothscroll' cursor is one screen line too far down. (Ernie
            Rael)
Solution:   Add a test that currently has the wrong result so that a fix can
            be made. (issue vim/vim#11436)

https://github.com/vim/vim/commit/75ac25b4967cdcdfdf2d6c086a6e2308868c280a

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0900: cursor moves too far with 'smoothscroll'

Problem:    Cursor moves too far with 'smoothscroll'.
Solution:   Only move as far as really needed. (Yee Cheng Chin, closes vim/vim#11504)

https://github.com/vim/vim/commit/81ba26e9de24ca6b1c05b6ec03e53b21793f1a4b

Co-authored-by: Yee Cheng Chin <ychin.git@gmail.com>

luukvbaal committed:
vim-patch:9.0.0901: setting w_leftcol and handling side effects is confusing

Problem:    Setting w_leftcol and handling side effects is confusing.
Solution:   Use a function to set w_leftcol() and handle side effects.

https://github.com/vim/vim/commit/0c34d562647f029faca40f7733ccfb7b5377672b

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.0908: with 'smoothscroll' cursor may end up in wrong position

Problem:    With 'smoothscroll' cursor may end up in wrong position.
Solution:   Correct the computation of screen lines. (Yee Cheng Chin,
            closes vim/vim#11502)

https://github.com/vim/vim/commit/361895d2a15b4b0bbbb4c009261eab5b3d69ebf1

Co-authored-by: Yee Cheng Chin <ychin.git@gmail.com>

luukvbaal committed:
vim-patch:9.0.0911: with 'smoothscroll' set mouse click position may be wrong

Problem:    With 'smoothscroll' set mouse click position may be wrong.
Solution:   Adjust computations for w_skipcol. (Yee Cheng Chin, closes vim/vim#11514)

https://github.com/vim/vim/commit/e6392b102151ec69fad232bcf00591230cef8e1c

Co-authored-by: Yee Cheng Chin <ychin.git@gmail.com>

luukvbaal committed:
vim-patch:9.0.0998: "gk" may reset skipcol when not needed

Problem:    "gk" may reset skipcol when not needed.
Solution:   Only reset skipcol if the cursor column is less.

https://github.com/vim/vim/commit/1b73edd9ee40aec400f3611f59823cec5fd1c489

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.1000: with 'smoothscroll' skipcol may be reset unnecessarily

Problem:    With 'smoothscroll' skipcol may be reset unnecessarily.
Solution:   Check the line does actually fit in the window.

https://github.com/vim/vim/commit/b21b8e9ed081a6ef6b6745fe65d219b3ac046c3b

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.1121: cursor positioning and display problems with 'smoothscroll'

Problem:    Cursor positioning and display problems with 'smoothscroll' and
            using "zt", "zb" or "zz".
Solution:   Adjust computations and conditions. (Yee Cheng Chin,
            closes vim/vim#11764)

https://github.com/vim/vim/commit/db4d88c2adfe8f8122341ac9d6cae27ef78451c8

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.1247: divide by zero with 'smoothscroll' set and a narrow window

Problem:    Divide by zero with 'smoothscroll' set and a narrow window.
Solution:   Bail out when the window is too narrow.

https://github.com/vim/vim/commit/870219c58c0804bdc55419b2e455c06ac715a835

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal commented on src/nvim/plines.c:
```diff
@@ -189,10 +189,11 @@ int plines_win_col(win_T *wp, linenr_T lnum, long column)
 /// @param[out] nextp    if not NULL, the line after a fold
 /// @param[out] foldedp  if not NULL, whether lnum is on a fold
 /// @param[in]  cache    whether to use the window's cache for folds
+/// @param[in]  winheight when true limit to window height
```
Added a `winheight` argument to pass down to `plines_win_nofill()`. This lets us re-use `plines_win_full()` in the newly added `plines_correct_topline()` that needs to first `adjust_plines_for_skipcol()` before limiting to window height itself.

luukvbaal commented on test/functional/legacy/scroll_opt_spec.lua:
```diff
@@ -0,0 +1,655 @@
+  -- Test_smoothscroll_zero_width
+  it("does not divide by zero with a narrow window", function()
+    screen:try_resize(60, 6)
+    -- FIXME: cannot split with 'winsize' "0 0" and does not crash without
```
Need to find a way to reproduce this failure and confirm that it's fixed.

luukvbaal reviewed (commented):
I think this is ready for general review. I marked out some of the changes that I made that didn't port one-to-one. I'll go over all the patches one more time to see if I missed any.

What's left to do:
* decide what to do with [this](https://github.com/neovim/neovim/pull/23320#issuecomment-1522743382) added vim9script test. I ported it to vimscript in-place for now.

zeertzjq commented on src/nvim/move.c:
```diff
@@ -1691,27 +2102,68 @@ void scroll_cursor_bot(int min_scroll, int set_topbot)
+
+    // If not using smoothscroll, we have to iteratively find how many
+    // lines to scroll down to roughly fit the cursor.
+    // This may not be right in the middle if the lines' physical height >
+    // 1 (e.g. 'wrap' is on).
+
```
Duplicate comment
```suggestion
```

luukvbaal replied:
Huh I guess it was added by accident when porting a different patch?

zeertzjq replied:
No, I added it intentionally because the position of the linebreak changed.

luukvbaal replied:
Do you want me to change the linebreak to stay compatible with vim? I may have changed the linebreak some other comments as well. I thought we allowed/preferred longer lines.

I guess the point is being able to `patch` without conflicts? I did everything by hand because most of it doesn't apply cleanly anyways because of ifdefs/formatting...

zeertzjq replied:
Hmm, I do think having a line break after `>` sign looks a bit strange.

luukvbaal replied:
Sure

luukvbaal replied:
> No, I added it intentionally because the position of the linebreak changed. The existing comment has the latest linebreak position.

I meant that this comment was originally added in a smoothscroll patch, so it was unexpected to already have in our code. Hence why I missed it and it ended up duplicated.

luukvbaal committed:
vim-patch:9.0.1367: divide by zero in zero-width window

Problem:    Divide by zero in zero-width window.
Solution:   Check the width is positive.

https://github.com/vim/vim/commit/e0f869196930ef5f25a0ac41c9215b09c9ce2d3c

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal committed:
vim-patch:9.0.1435: scrolling too many lines when 'wrap' and 'diff' are set

Problem:    Scrolling too many lines when 'wrap' and 'diff' are set.
Solution:   Only scroll by screenlines for 'diff' when 'wrap' is not set.
            (closes vim/vim#12211)

https://github.com/vim/vim/commit/38d867f041349e1400c2cce9cac06f59ae6ccbb1

Co-authored-by: Bram Moolenaar <Bram@vim.org>

luukvbaal commented:
> vim-patch:9.0.1247: oldtest divide by zero does not work ported to functionaltest
find a way to make this fail pre v9.0.1247 in neovim and port to functionaltest. this test doesn't fail with vim either on v9.0.1246 so IDK

I think I updated the test to behave the same as vim (which does not abort the screendump test on errors) by adding `sil!`. However neither the functionaltest for us nor the oldtest for vim fail on `v9.0.1246` so the test seems pointless.

luukvbaal committed:
test: 'smoothscroll' works with virt_lines above and below

luukvbaal committed:
vim-patch:9.0.1502: no test for deleting the end of a long wrapped line

Problem:    No test for deleting the end of a long wrapped line.
Solution:   Add a test to check the right text is displayed. (Luuk van Baal,
            closes vim/vim#12318)

https://github.com/vim/vim/commit/5b10a140983c16140d69a214494c4b8af8b34763

Co-authored-by: Luuk van Baal <luukvbaal@gmail.com>

luukvbaal commented:
I went over all patch diffs side by side one more time, taking extra care to make sure that the ported screendump tests are identical.
Ready to merge AFAIAC:shipit:

zeertzjq commented on src/nvim/grid.c:
```diff
@@ -529,6 +529,30 @@ void grid_put_linebuf(ScreenGrid *grid, int row, int coloff, int endcol, int cle
   max_off_from = linebuf_size;
   max_off_to = grid->line_offset[row] + (size_t)grid->cols;
 
+  // Take care of putting "<<<" on the first line for 'smoothscroll'.
```
This doesn't work with tabline, winbar, or splits above.

zeertzjq commented on src/nvim/grid.c:
```diff
@@ -529,6 +529,30 @@ void grid_put_linebuf(ScreenGrid *grid, int row, int coloff, int endcol, int cle
   max_off_from = linebuf_size;
   max_off_to = grid->line_offset[row] + (size_t)grid->cols;
 
+  // Take care of putting "<<<" on the first line for 'smoothscroll'.
+  if (row == 0 && wp->w_skipcol > 0
+      // do not overwrite the 'showbreak' text with "<<<"
+      && *get_showbreak_value(wp) == NUL
+      // do not overwrite the 'listchars' "precedes" text with "<<<"
+      && !(wp->w_p_list && wp->w_p_lcs_chars.prec != 0)) {
+    int off = 0;
+    int skip = 0;
+    if (wp->w_p_nu && wp->w_p_rnu) {
+      // do not overwrite the line number, change "123 text" to
+      // "123>>>xt".
+      while (skip < wp->w_width && ascii_isdigit(*linebuf_char[off])) {
+        off++;
+        skip++;
+      }
+    }
+
+    for (int i = 0; i < 3 && i + skip < wp->w_width; i++) {
+      schar_from_ascii(linebuf_char[off], '<');
+      linebuf_attr[off] = HL_ATTR(HLF_AT);
+      off++;
+    }
+  }
+
```
This doesn't work with tabline, winbar, or splits above.

luukvbaal replied:
Right, how about this:
```diff
diff --git a/src/nvim/grid.c b/src/nvim/grid.c
index f02ff7951..2c548e11f 100644
--- a/src/nvim/grid.c
+++ b/src/nvim/grid.c
@@ -503,6 +503,7 @@ void grid_put_linebuf(ScreenGrid *grid, int row, int coloff, int endcol, int cle
   int col = 0;
   bool redraw_next;                         // redraw_this for next character
   bool clear_next = false;
+  bool topline = row == 0;
   int char_cells;                           // 1: normal char
                                             // 2: occupies two display cells
   int start_dirty = -1, end_dirty = 0;
@@ -529,7 +530,7 @@ void grid_put_linebuf(ScreenGrid *grid, int row, int coloff, int endcol, int cle
   max_off_from = linebuf_size;
   max_off_to = grid->line_offset[row] + (size_t)grid->cols;

-  if (row == 0 && wp->w_skipcol > 0) {
+  if (topline && wp->w_skipcol > 0) {
     // Take care of putting "<<<" on the first line for 'smoothscroll'.
     for (int i = 0; i < 3; i++) {
       schar_from_ascii(linebuf_char[i], '<');
```

luukvbaal committed:
test: 'smoothscroll' <<< marker shows with tabline, winbar and splits

zeertzjq commented on runtime/doc/options.txt:
```diff
@@ -5671,6 +5671,16 @@ A jump table for the options with a short description can be found at |Q_op|.
 	The value of 'softtabstop' will be ignored if |'varsofttabstop'| is set
 	to anything other than an empty string.
 
+			*'smoothscroll'* *'sms'* *'nosmoothscroll'* *'nosms'*
+'smoothscroll' 'sms'	boolean  (default off)
+			local to window
+	Scrolling works with screen lines.  When 'wrap' is set and the first
+	line in the window wraps part of it may not be visible, as if it is
+	above the window. "<<<" is displayed at the start of the first line,
+	highlighted with |hl-NonText|.
+	NOTE: only partly implemented, currently works with CTRL-E, CTRL-Y
+	and scrolling with the mouse.
+
```
This should be before 'softtabstop', not after.

zeertzjq commented on src/nvim/buffer_defs.h:
```diff
@@ -193,6 +193,8 @@ typedef struct {
   long wo_scr;
 #define w_p_scr w_onebuf_opt.wo_scr     // 'scroll'
   int wo_spell;
+#define w_p_sms w_onebuf_opt.wo_sms     // 'smoothscroll'
+  int wo_sms;
```
```suggestion
  int wo_sms;
#define w_p_sms w_onebuf_opt.wo_sms     // 'smoothscroll'
  int wo_spell;
```

zeertzjq commented:
The commit message of v9.0.0641 should be squashed into v9.0.0640.

zeertzjq commented on test/old/testdir/test_options.vim:
```diff
@@ -775,6 +775,108 @@ func Test_copy_winopt()
+    let values += [
+      \['cursorcolumn', 1, 0],
+      \['cursorline', 1, 0],
+      \['colorcolumn', '"+1"', '"' .. &colorcolumn .. '"'],
+      \['cursorlineopt', '"screenline"', '"' .. &cursorlineopt .. '"'],
```
```suggestion
      \['cursorlineopt', '"screenline"', '"' .. &cursorlineopt .. '"'],
      \['colorcolumn', '"+1"', '"' .. &colorcolumn .. '"'],
```

zeertzjq commented on test/functional/legacy/scroll_opt_spec.lua:
```diff
@@ -0,0 +1,776 @@
+      with some text with some text           |
+                                              |
+    ]])
+    feed('2j02k')
```
```suggestion
    -- moving cursor up where the >>> marker is - whole top line shows
    feed('2j02k')
```

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -2890,27 +2893,21 @@ static void nv_zet(cmdarg_T *cap)
     FALLTHROUGH;
 
-  // "zl" - scroll screen to the left
+  // "zl" - scroll window to the left
```
```suggestion
  // "zl" - scroll window to the left if not wrapping
```

zeertzjq commented on test/old/testdir/test_window_cmd.vim:
```diff
@@ -1889,7 +1889,25 @@ function Test_new_help_window_on_error()
+  set cpo-=n
+  unmap 0
+  set nonumber nosmoothscroll
+endfunc
```
Missing blank line
```suggestion
endfunc

```

zeertzjq commented:
LGTM for non-multigrid UI. For multigrid UI most of these `w_width` and `w_height` should need to be changed `w_width_inner` and `w_height_inner`.

luukvbaal commented:
Thanks for the review, I'll include the changes.

> For multigrid UI most of these w_width and w_height should need to be changed to w_width_inner and w_height_inner.

Yeah I was wondering about this when porting the patches but noticed other uses of `w_height/width` in `move.c` still. Should every occurrence simply be replaced with the inner variant?

luukvbaal commented:
> For multigrid UI most of these w_width and w_height should need to be changed to w_width_inner and w_height_inner.

> > Yeah I was wondering about this when porting the patches but noticed other uses of w_height/width in move.c still. Should every occurrence simply be replaced with the inner variant?

Should we `if multrigrid else` wrap the 'smoothscroll' tests and verify by hand?

zeertzjq commented:
Yes I think all window widths and heights in this PR should be the inner variant. Only running the tests with ext_multigrid isn't really useful as `w_width_inner` is still the same as `w_height_inner` without calling `nvim_ui_try_resize_grid`.

luukvbaal committed:
fix(ui): adjust 'smoothscroll' for inner dimensions

zeertzjq commented:
There are two `w_height` in `adjust_skipcol()` and `pline_correct_topline()` that should be changed to `w_height_inner`.

zeertzjq commented:
Also, in the first commit, co-author isn't recognized if it isn't at the end of the commit message. And if you rebase on master the two `w_height` mentioned above should be the only remaining occurrences of `w_width` or `w_height` in `move.c`.

rishabh-ranjan commented:
hi, thanks for the feature. when will this be available in an official release? can't find it in the latest (v0.9.4)
